### DISCUSSION TITLE
Working with univariate polynomial ideals?

### DISCUSSION THREAD

**Discussion by oskarhenriksson** [oscar-system/Oscar.jl#3504] [Q&A] [answered]
Some of my applied algebra students are a bit confused about how to deal with the univariate polynomial ring $\mathbb{Q}[x]$ in Oscar. The issue is that there are two different ways to think about them, corresponding to two different data types in Oscar:
 - truly univariate polynomial rings of type `QQPolyRing` 
 (obtained by writing `R,x=polynomial_ring(QQ,"x")` or `R,x=QQ["x"]`)
 - multivariate polynomial rings of type `QQMPolyRing` that happen to just have a single generator 
 (obtained via `R,(x,)=polynomial_ring(QQ,["x"])`).

You need `QQPolyRing` to be able to do tasks that are specific to the univariate case (like computing `roots`), but you need ``QQMPolyRing`` to do general computations for polynomial rings (like `groebner_basis` or `primary_decomposition`). 

If you need both points of view, you need to create two ideals in two rings, which easily gets confusing in a teaching situation...

Is there any way to make this more beginner-friendly? Naively, it would of course be nice if as much as possible of the `QQMPolyRing` functionality just automatically worked on `QQPolyRing` objects, but perhaps `QQPolyRing` plays roles in other parts of Oscar that would make this undesirable? (For instance, `ideal(R,[x^2+2])` gives an ideal of type `Hecke.PIDIdeal{QQPolyRingElem}`, so this is related to the Hecke algebra parts.)

**Comment by fieker** [comment, marked as answer]
There are several things, but I guess some would be (can be) highly confusing. The most important difference, from a proactical point of view, is the different interface to the internals (coefficients, ...). Univariate are always dense, multivariate types are sparse.
Thus for `f = x^3-2`, a call to `coeff(f, 0)` will give -2 in the univariate case and an error in the multivariate. In the multivariate case, `coeff(f, 1) -> 1`, `coeff(f, 2) -> -2` and all others error. This allows to implement different algorithms for arithmetic
 - univariate multiplication of 2 degree n polynomials can be done in O(n log(n) operations
  - multivariate of 2 polynomials with n terms is O(n^2)
  
 However that means in the univariate case `(x^100-1) *(x^101-1)` will take > 100 operations, while its 4 in the multivariate case.

There is no reason why we cannot add roots to multivariates - after all we can already factor... that would be trivial.
As for ideals: univariate ideals were only added as people wanted them, not due to any (serious) application. Having said that, univariate rings (over fields) have a simple structure that we want to exploit Also, my vague recollection: the gcd(f, g) is a Groebner basis to the ideal generated by f and g...

So maybe the more intersting question: what are your applications? What would help you? (roots is easy:
```
Oscar.roots(f::MPolyRingElem) = [-constant_coefficient(t)//leading_coefficient(t) for (t, e) = factor(f) if total_degree(t) == 1]
```
OK, this is not optimal...)

**Comment by thofma** [comment]
I think it would be great, if the ideal functionality available for the `*MPoly` rings would also work for the univariate rings. There is unfortunately nothing automatic we could do. We add them as they are needed/requested. Apart from `groebner_basis` and `primary_decomposition`, are there any other functions on your list?
